File: src/shared/admin/items.ts
import { getEnabledProviders, getProvider, type DashboardItem, type ItemDetail } from "../../core/registry.js";

export function listItems(): DashboardItem[] {
  const items = getEnabledProviders().flatMap((p) => p.listItems?.() ?? []);
  return items.sort((a, b) => (a.createdAt < b.createdAt ? 1 : -1));
}

export function getItemDetail(provider: string, id: string): ItemDetail | undefined {
  return getProvider(provider)?.getItemDetail?.(id);
}

File: internal/shared/admin/items.go
package admin

import (
	"sort"

	"github.com/golgimed/mimic/internal/registry"
)

// ListItems aggregates listItems() across every enabled provider, newest first.
func ListItems(reg *registry.Registry, enabledProviders []string) []registry.DashboardItem {
	items := []registry.DashboardItem{}
	for _, p := range reg.Enabled(enabledProviders) {
		if p.ListItems == nil {
			continue
		}
		items = append(items, p.ListItems()...)
	}
	sort.SliceStable(items, func(i, j int) bool {
		return items[i].CreatedAt > items[j].CreatedAt
	})
	return items
}

func GetItemDetail(reg *registry.Registry, provider, id string) (*registry.ItemDetail, bool) {
	p, ok := reg.Get(provider)
	if !ok || p.GetItemDetail == nil {
		return nil, false
	}
	return p.GetItemDetail(id)
}
